add support for proxy auth flow and secured S3-backed registries

jtriley opened the pull request:
**Description of the Pull Request (PR):**

NVIDIA cloud's private docker registry flow is slightly different than the flow used in singularity's python docker registry API and is supported by the official docker client. In addition to a slightly different flow, NVIDIA cloud's private docker registry is also backed by a secured S3 bucket which singularity currently doesn't handle properly. This PR fixes both of these issues and allows converting images from NVIDIA cloud's private docker registry into singularity images.

With the proxy auth challenge only the realm is returned and the only requirement is to auth against that realm URL (see singularityware/singularity#1159 for more details). To address this the challenge parsing regex and code has been updated to handle any combination of realm, service, and scope (as long as that order is maintained) and builds the token URL accordingly. **NOTE**: to be fully robust in our challenge parsing we'd need to use something like https://github.com/alexsdutton/www-authenticate but for now this is likely close enough for most docker registry use cases.

Once you get past the auth challenge parsing and token URL generation issue the next hurdle is that fetching layers involves a redirect to pre-signed S3 urls. Currently when this happens the auth credentials are passed along with the request which causes an InvalidArgument error about using more than one auth mechanism due to the pre-signed URL already containing an `X-Amz-Algorithm` query param and singularity additionally providing the `Authorization:` header with the bearer token (see singularityware/singularity#1159 for more details). This PR employs an `HTTPHandler` and `HTTPSHandler` with urllib to intercept all requests, inspect the URL for a pre-signed S3 URL, and if present strips any `Authorization:` header from the request. This should fix things for any private registry using a secured S3-backend - not just NVIDIA's.

This approach ensures that we're only stripping the `Authorization:` header in the case of S3-backed docker registries that use pre-signed S3 URLS where this is known be an issue (e.g. NVIDIA's private registry: `nvcr.io`). In all other cases the previous behavior should be preserved.

**This fixes or addresses the following GitHub issues:**

- Ref: #1159, #1183

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

jtriley committed:
add support for proxy auth bearer challenge

Update challenge parsing to support private registries similar to
NVIDIA's private registry nvcr.io which returns bearer realm only.

See singularityware/singularity#1159 for more details

jtriley committed:
strip authorization header for pre-signed S3 urls

Without this we hit an InvalidArgument error about using more than
one auth mechanism for S3-backed private registries.

This approach uses an HTTPHandler/HTTPSHandler with urllib to intercept
all http/https requests, inspect the request URL for a pre-signed S3
URL, and if present strips any Authorization header from the request.

This approach ensures that we're only stripping the authorization header
for S3-backed docker registries that use pre-signed S3 URLS where this
is known be an issue (e.g. NVIDIA's private registry: nvcr.io). In all
other cases the previous behavior should be preserved.

See singularityware/singularity#1159 for more details

jtriley committed:
update CONTRIBUTORS.md

jtriley committed:
add bot.exception() for logging exceptions

Takes a message and appends the current traceback string and emits the
result at DEBUG level.

jtriley committed:
log all uncaught exceptions in download_atomically

...these get emitted at DEBUG log level.

yarikoptic commented on libexec/python/base.py:
```diff
@@ -126,6 +128,51 @@ def multi_package(func, args):
+    parts = urlparse(req.get_full_url())
+    host = parts.netloc
+    query_dict = parse_qs(parts.query)
+    is_s3_url = 's3' in host and host.endswith('.amazonaws.com')
```
minor nit picking, but where s3 would appear, may be at least `'s3' in host.split('.')` would avoid matching some possibly unrelated urls (I don't know if any exist but I guess smth like `auths3.amazonaws.com` could emerge at some point)

vsoch replied:
I do think that these edge cases are likely to occur (eventually) but we can address a more general strategy with subsequent PRs to this very specific use case - it would include more thinking about "all the possible versions of s3" and other storage providers generally, along with auth schemas and challenges. The way I'm thinking about it now is that this PR makes the software more functional than without, and is a step forward.

jtriley replied:
@yarikoptic Unfortunately the URL varies with location/availability zone so we can't just look for `'s3' in host.split('.')`. Typically the URL starts with two sub-domains: the bucket name and a location-specific string starting with `s3`, e.g.:

```
https://some-bucket-name.s3-us-west-1.amazonaws.com
```

So the current code looks for s3 anywhere in the string before the `.amazonaws.com` domain. The only update that would make this a bit more S3-specific without maintaining a list of availability zones in AWS is to instead check that first location-specific sub-domain for something that matches `^s3-.*$`. @vsoch Let me know if you'd prefer to have that update or leave things as-is.

vsoch replied:
I think given that we can expect to later make this more elegant, it would suffice as is, but @yarikoptic could you add your thoughts if this update would address some of your concern? There will always be "one more pattern" missed by a single regular expression, and when we have classes for each set it will be more manageable.

jtriley replied:
After some research it appears there are several variations depending on whether you're using virtual-host or path style bucket URLS:

https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro

And there's another one that's not listed in that doc according to this issue from another repo:

https://github.com/zalando/PGObserver/issues/91

I'll work on a regex that covers all those cases but there's still the possibility of other variants out there and the current code is pretty flexible as-is.

yarikoptic reviewed (commented):
didn't have a chance to try but had a chance to look through.  minor comment only

jtriley committed:
base.py: simplify AWS/S3 url detection

The pre-signed URL headers are specific to AWS/S3 and known to throw an
error with an 'Authorization' header present. Rather than inspect the
host in the request just strip the auth header any time these
s3-specific headers are present. This is much easier and more flexible
than hard-coding AWS/S3 detection via regex matching based on the
hostname of the url in the request. Should also allow alternate S3
implementations/deployments to work (e.g. Swift).

dtrudg reviewed (approved):
@jtriley - this looks good to me as a way of getting this handled in the 2.x codebase. Has tested okay for me with nvcr.io and multiple containers. I can sort out where exactly it needs to go / target branch, given the 2.4.x 2.5 discussion yesterday. Thanks!
